feat(docagent): hypothetical questions

dfm88 opened the pull request:
### Hypothetical Question Index

Despite there being no papers supporting this technique, Hypothetical Question Indexing is mentioned in several blogs as an advanced RAG technique. It somewhat contrasts with Hypothetical Answer (HyDE) and intervenes during the embedding saving phase, specifically trying to solve the problem where the question posed on an embedding contains words distant from it. Therefore, it associates different questions to increase the probability of retrieval.

Some link:
- https://www.falkordb.com/blog/advanced-rag/
- https://youtu.be/sGvXO7CVwc0?t=158

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -405,6 +419,10 @@ def ingest_docs(
             raise ValueError("VecDB not set")
 
+        if self.config.hypothetical_questions:
+            question_docs = [self.llm_hypothetical_question(doc) for doc in docs]
```
This will be very slow when there are hundreds/thousands of chunks. We should use an appropriate batch method, see `langroid/agent/batch.py`

dfm88 replied:
done

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -126,6 +126,20 @@ class DocChatAgentConfig(ChatAgentConfig):
+    # which uses a language model to generate questions for each data chunk,
+    # this might help in finding relevant chunks for a given question.
+    hypothetical_questions: bool = False
+    hypothetical_questions_nr: int = 3
```
=> `num_hypothetical_questions`

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -126,6 +126,20 @@ class DocChatAgentConfig(ChatAgentConfig):
+    # this might help in finding relevant chunks for a given question.
+    hypothetical_questions: bool = False
+    hypothetical_questions_nr: int = 3
+    hypothetical_questions_default_prompt: str = f"""
```
=> `hypothetical_questions_prompt`

dfm88 replied:
done

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -860,6 +878,26 @@ def llm_hypothetical_answer(self, query: str) -> str:
+                }
+            questions = self.llm_response_forget(prompt).content
+        return Document(
+            content=questions,
```
The content is just the questions, which is fine since it helps match the query, but we lose the original content. I know you are keeping it in the metadata, but `original_content` is not used in the normal flow of this agent anywhere. 

I think it's better to _augment_ the `content` field value with the `questions`, and also demarcate those questions in some special way, so that during passage retrieval, the questions are ignored. I.e. only use the questions for matching/ranking, but ignore in retrieval.

pchalasani replied:
The other advantage of augmenting (i.e. combining chunk content + questions) is that now we can match with both the question and content, rather than having to deal with de-duplication, e.g. when we find a match for questions_5 that links to chunk_3 and also find a match to chunk_3.

dfm88 replied:
Thanks for the feedback! 

I've implemented the suggested improvements like this:

1. Combined content+questions in a single field instead of separating them, using a consistent marker (`GENERATED_QUESTIONS_MARKER`) to demarcate them.

2. Added `clean_generated_content()` method in the so that questions are used in ranking but not showed in retrieval

pchalasani replied:
Thank you @dfm88 ! Added to my queue of things to look at this week.

pchalasani commented:
Thanks for this @dfm88 
Please also add suitable tests in `test_doc_chat_agent.py`

dfm88 commented on langroid/agent/chat_agent.py:
Here I had to change the signature of 
`llm_response_forget` and `llm_response_forget_async` (which might also have sense) because I had MyPy problem in passing the `llm_response_forget_async` method to `run_batch_agent_method`

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -860,6 +879,60 @@ def llm_hypothetical_answer(self, query: str) -> str:
+
+        with status("[cyan]Generating hypothetical questions for chunks..."):
+            # Process chunks in parallel using run_batch_agent_method
+            questions_batch = run_batch_agent_method(
```
there could be 1000+ chunks so we should use a batched version of this method. I know there's no `batch_size` param in this method, so we'd have to add that to `run_agent_method_batch` and related onces

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -1315,6 +1388,35 @@ def get_relevant_extracts(self, query: str) -> Tuple[str, List[Document]]:
 
         return query, extracts
 
+    def clean_generated_content(self, passages: List[Document]) -> List[Document]:
```
suggest rename - `remove_hypothetical_questions` to be more clear

dfm88 replied:
I thoughy it might need for other cleaning stuff so I used a generic name, but ok I'll make it specific for this task

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -126,6 +128,20 @@ class DocChatAgentConfig(ChatAgentConfig):
+    # Flag to enable Hypothetical Question Indexing,
+    # which uses a language model to generate questions for each data chunk,
+    # this might help in finding relevant chunks for a given question.
+    hypothetical_questions: bool = False
```
add config:
hypothetical_questions_batch_size: int = 50

pchalasani commented on tests/main/test_doc_chat_agent.py:
```diff
@@ -939,3 +940,165 @@ def gen_task(i: int):
+@pytest.mark.parametrize(
+    "hypothetical_questions, expect_cleaned", [(True, True), (False, False)]
+)
+def test_clean_generated_content(
```
==> `test_remove_generated_questions`

pchalasani commented on tests/main/test_doc_chat_agent.py:
It would be nice if you can think of a small test that shows the benefit of adding the hyp questions.

Or better, an example scripts under `examples/docqa/` would be great.

pchalasani reviewed (changes requested):
Thanks @dfm88 I just reviewed it.

dfm88 commented:
Hi @pchalasani , sorry for the delay in making changes on this. Here an overview of the changes, since it was not a straightforward task to handle batch number:

I saw you added a function in `batch.py` file to handle batches, that actually was handling only sync code, I slughtly changed it to also handle aysnc code and try to reuse it in the `run_batch_agent_method` 

The most challenging part was ensuring correct behavior when stop_on_first_result=Truem so I maintained strict task ordering by preserving original indices.
Force batch_size=1 when stop_on_first_result=True to prevent later tasks from completing first
Use an async lock to ensure thread-safe access when storing the first completed result
Carefully reconstruct the result list to preserve the original task positions

For normal operation (when stop_on_first_result=False), the code fully leverages the batching capabilities.

I have also added an example of when leveraging Hypothetical Questions might be beneficial. Here another interesting article [Advanced RAG Optimization: Aligning Question and Document Embedding Spaces with Hypothetical Questions](https://blog.epsilla.com/demystifying-rag-empowered-chat-agents-aligning-question-and-document-embedding-spaces-with-5710c4218464)

I know there are a lot of changes, but tests using batch module are passing. Let me know if you'd like me to clarify any part of these changes or have suggestions for improvement.

pchalasani commented:
Thank you @dfm88 will check in a few days.

pchalasani commented on langroid/agent/batch.py:
```diff
@@ -281,50 +297,64 @@ def run_batch_agent_method(
 
-    async def _do_task(input: str | ChatDocument, i: int) -> Any:
-        agent_cfg.name = f"{agent_cfg.name}-{i}"
+    first_result = None
```
Thanks for this @dfm88.

Is there a reason the implementation of `stop_on_first_result` here is different from the one in `run_batch_task_gen` -- is this somehow better? This will help me better understand + review.

dfm88 replied:
Hi @pchalasani,

thank you for pointing this out. I actually didn't notice the `run_batch_task_gen` implementation when I was working on this PR. Looking at it now, I can see it already provides a good handling of batch async tasks, so I basically took inspiration from that to implement `batch_size` also in `run_batch_agent_method` in my last commit.

So I could revert also the changes I did to `langroid/agent/batch.py::run_batch_function`.

Finally I fixed the tests where I still was using the previous naming for the doc agent method relative to hypothetical questions.

pchalasani replied:
Will review soon thanks !

pchalasani commented on langroid/agent/batch.py:
```diff
@@ -312,17 +317,40 @@ async def _do_all() -> List[Any]:
+                f"{agent_name}{complete_str}..."
+            )
+
+            if stop_on_first_result and any(results):
```
I think we should discard None results here, just like we do in the `run_batch_task_gen` fn.

Also more generally, since this `_do_all` functions in the two places are so similar, I think we should avoid code duplication and have a common fn and use it in both places. What do you think @dfm88 ?

pchalasani commented on examples/docqa/hypothetical_questions.py:
```diff
@@ -0,0 +1,154 @@
+    doc_task = Task(
+        doc_agent,
+        interactive=False,
+        name="DocAgent",
```
set `single_round=True` so it exits after generating the answer -- this avoids the current behavior of "stuck for 5 steps"

pchalasani commented on examples/docqa/hypothetical_questions.py:
I ran this with and without hyp questions and got the exact same  result:

<img width="1442" alt="image" src="https://github.com/user-attachments/assets/f5882272-b366-4fa4-bc40-54e1034e3777" />

dfm88 replied:
Oh yeah, I see. So given that Hypothetical Question works well in particular scenarios, (e.g., specialized jargon, very technical documents, Q-A) I provided a new example on a langorid python file (doc_chat_agent.py -Inception?-), and the user query `"How does the system generate responses in conversation mode?"` got a more appropriate answer using HQ.

Without HQ:
![WITHOUT_HQ](https://github.com/user-attachments/assets/8d5973ef-4409-4473-89ff-fc6a69045ed3)

With HQ
![WITH_HQ](https://github.com/user-attachments/assets/f0a7db4c-1af7-4899-9f29-3ead3e5117e6)

pchalasani replied:
> doc_chat_agent.py -Inception?-

Didn't understand what you meant by `inception` ?

pchalasani commented on langroid/agent/special/doc_chat_agent.py:
```diff
@@ -81,6 +82,8 @@ def apply_nest_asyncio() -> None:
 You are a helpful assistant, helping me understand a collection of documents.
 """
 
+GENERATED_QUESTIONS_MARKER = "--- Generated Questions ---"
```
The chunk-embeddings will be distracted by the presence of the words "generated questions", and it would thus affect retrieval as well. Instead, this should be set to something non-semantic but unique like 
```
<##-##-##>
```

To be even safer, we could try to make the vecdb aware of the presence of this marker and strip it out before embedding, but that is messy (we have to propagate the hypo questions flag all the way into the `VectorStoreConfig` -- we can skip that for now.

pchalasani commented:
@dfm88 this is going to be very useful, as you say in cases with very technical/domain-specific terms that are not captured well by generic embeddings. I changed your example script to the medical domain to illustrate the benefit more clearly and explicitly, see the `docagent-hypoq` branch:
https://github.com/langroid/langroid/blob/docagent-hypoq/examples/docqa/hypothetical_questions.py

I added explanatory notes at the top of the script.

NOTE - you need to pull latest from main (there is a minor fix in `doc_chat_agent.py` and a needed method in `mytypes.py`) and then pull the `hypothetical_questions.py` script from the `docagent-hypoq` branch.

dfm88 commented:
> @dfm88 this is going to be very useful, as you say in cases with very technical/domain-specific terms that are not captured well by generic embeddings. I changed your example script to the medical domain to illustrate the benefit more clearly and explicitly, see the `docagent-hypoq` branch: https://github.com/langroid/langroid/blob/docagent-hypoq/examples/docqa/hypothetical_questions.py
> 
> I added explanatory notes at the top of the script.
> 
> EDIT - Please merge the changes from branch `docagent-hypoq` into your branch. The `docagent-hypoq` branch also includes some fixes to `doc_chat_agent.py` and a new method `Document.from_string` in `mytypes.py`

Great  thanks! I did the merge

pchalasani commented:
@dfm88 I am making some changes on top of this.

dfm88 commented:
> @dfm88 I've made a new PR with my changes on top, #683 I made the design more flexible, to support any type of chunk enrichment, and also allow a different LLM to be used for the chunk enrichment. Essentially, we can can pass a config for a "chunk enrichment agent".

Yeah I see, it makes a lot of sense to me, thanks!

pchalasani commented:
> Yeah I see, it makes a lot of sense to me, thanks!

Thank you for leading this PR !

pchalasani commented:
finished in #683
